RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc

```diff
@@ -0,0 +1,974 @@
+  }
+}
+
+//
+// Produce seeds
+//
+void TSGForOIDNN::produce(edm::StreamID sid, edm::Event& iEvent, edm::EventSetup const& iEventSetup) const {
+  // Initialize variables
+  unsigned int numSeedsMade = 0;
+  unsigned int layerCount = 0;
+  unsigned int hitlessSeedsMadeIP = 0;
+  unsigned int hitlessSeedsMadeMuS = 0;
+  unsigned int hitSeedsMade = 0;
+  unsigned int hitDoubletSeedsMade = 0;
+
+  // Surface used to make a TSOS at the PCA to the beamline
+  Plane::PlanePointer dummyPlane = Plane::build(Plane::PositionType(), Plane::RotationType());
+
+  // Read ESHandles
+  edm::ESHandle<Chi2MeasurementEstimatorBase> estimatorH = iEventSetup.getHandle(t_estimatorH_);
+  edm::ESHandle<MagneticField> magfieldH = iEventSetup.getHandle(t_magfieldH_);
+  edm::ESHandle<Propagator> propagatorAlongH = iEventSetup.getHandle(t_propagatorAlongH_);
+  edm::ESHandle<Propagator> propagatorOppositeH = iEventSetup.getHandle(t_propagatorOppositeH_);
+  edm::ESHandle<TrackerGeometry> tmpTkGeometryH = iEventSetup.getHandle(t_tmpTkGeometryH_);
+  edm::ESHandle<GlobalTrackingGeometry> geometryH = iEventSetup.getHandle(t_geometryH_);
+  edm::ESHandle<NavigationSchool> navSchool = iEventSetup.getHandle(t_navSchool_);
+
+  edm::Handle<MeasurementTrackerEvent> measurementTrackerH;
+  iEvent.getByToken(measurementTrackerTag_, measurementTrackerH);
+
+  // Read L2 track collection
+  edm::Handle<reco::TrackCollection> l2TrackCol;
+  iEvent.getByToken(src_, l2TrackCol);
+
+  // The product
+  std::unique_ptr<std::vector<TrajectorySeed> > result(new std::vector<TrajectorySeed>());
+
+  // Get vector of Detector layers
+  std::vector<BarrelDetLayer const*> const& tob = measurementTrackerH->geometricSearchTracker()->tobLayers();
+  std::vector<ForwardDetLayer const*> const& tecPositive =
+      tmpTkGeometryH->isThere(GeomDetEnumerators::P2OTEC)
+          ? measurementTrackerH->geometricSearchTracker()->posTidLayers()
+          : measurementTrackerH->geometricSearchTracker()->posTecLayers();
+  std::vector<ForwardDetLayer const*> const& tecNegative =
+      tmpTkGeometryH->isThere(GeomDetEnumerators::P2OTEC)
+          ? measurementTrackerH->geometricSearchTracker()->negTidLayers()
+          : measurementTrackerH->geometricSearchTracker()->negTecLayers();
+
+  // Get suitable propagators
+  std::unique_ptr<Propagator> propagatorAlong = SetPropagationDirection(*propagatorAlongH, alongMomentum);
+  std::unique_ptr<Propagator> propagatorOpposite = SetPropagationDirection(*propagatorOppositeH, oppositeToMomentum);
+
+  // Stepping Helix Propagator for propogation from muon system to tracker
+  edm::ESHandle<Propagator> SHPOpposite = iEventSetup.getHandle(t_SHPOpposite_);
+
+  // Loop over the L2's and make seeds for all of them
+  LogTrace(theCategory_) << "TSGForOIDNN::produce: Number of L2's: " << l2TrackCol->size();
+  for (unsigned int l2TrackColIndex(0); l2TrackColIndex != l2TrackCol->size(); ++l2TrackColIndex) {
+    const reco::TrackRef l2(l2TrackCol, l2TrackColIndex);
```

Comment:
```suggestion
  for (auto const& l2 : l2TrackCol) {
```
the ref is apparently not needed in this module; it is wasteful to create a Ref so that it can be dereferenced back

Reply:
Done!